src/runtime/HalideRuntime.h

```diff
@@ -541,6 +546,38 @@ extern int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf);
 extern int halide_copy_to_device(void *user_context, struct halide_buffer_t *buf,
                                  const struct halide_device_interface_t *device_interface);
 
+/** Copy data from one buffer to another. The buffers may have
+ * different shapes and sizes, but the destination buffer's shape must
+ * be contained within the source buffer's shape. The source data is
+ * pulled from either device or host memory on the source, depending
+ * on the dirty flags. host is preferred if both are valid. The
+ * dst_device_interface parameter controls the destination memory
+ * space. NULL means host memory. */
```

Comment:
We should also probably state explicitly whether halide_buffer_copy and halide_buffer_crop are allowed to modify the host memory in device to device cases and such.

It might also be useful to model halide_copy_to_host and halide_copy_to_device in terms of halide_buffer_copy and provide that mapping for illustrative purposes.